apps/site/pages/en/learn/test-runner/mocking.md

```diff
@@ -0,0 +1,264 @@
+
+```mjs displayName="storage.mjs"
+import { db } from 'db';
+
+export function read(key, all = false) {
+  validate(key, val);
+
+  if (all) return db.getAll(key);
+
+  return db.getOne(key);
+}
+
+export function save(key, val) {
+  validate(key, val);
+
+  return db.upsert(key, val);
+}
+```
+
+```mjs displayName="storage.test.mjs"
+import assert from 'node:assert/strict';
+import { describe, it } from 'node:test';
+
+import { db } from 'db';
+
+import { save } from './storage.mjs';
+
+describe('storage', { concurrency: true }, () => {
+  it('should retrieve the requested item', async () => {
+    await db.upsert('good', 'item'); // give it something to read
+    await db.upsert('erroneous', 'item'); // give it a chance to fail
+
+    const results = await read('a', true);
+
+    assert.equal(items.length, 1); // ensure read did not retrieve erroneous item
+
+    assert.deepEqual(results[0], { key: 'good', val: 'item' });
+  });
+
+  it('should save the new item', async () => {
+    const id = await save('good', 'item');
+
+    assert.ok(id);
+
+    const items = await db.getAll();
+
+    assert.equal(items.length, 1); // ensure save did not create duplicates
+  });
+});
+```
+
+In the above, the first and second cases (the `it`s) can sabotage each other because they are run concurrently and mutate the same store (a race condition): "save"'s insertion can cause the otherwise valid "read"'s test to fail its assertion on items found (and "read"'s can do the same thing to "save"'s).
+
+## What to mock
+
+### Modules + units
+
+```mjs
+import assert from 'node:assert/strict';
+import { before, describe, it, mock } from 'node:test';
+
+
+describe('foo', { concurrency: true }, () => {
+  let barMock = mock.fn();
+  let foo;
+
+  before(async () => {
+    const barNamedExports = await import('./bar.mjs')
+      // discard the original default export
+      .then(({ default, ...rest }) => rest);
+
+    // It's usually not necessary to manually call restore() after each
+    // nor reset() after all (node does this automatically).
+    mock.module('./bar.mjs', {
+      defaultExport: barMock
+      // Keep the other exports that you don't want to mock.
+      namedExports: barNamedExports,
+    });
+
+    // This MUST be a dynamic import because that is the only way to ensure the
+    // import starts after the mock has been set up.
+    // There is a far more technical explanation,
+    // but just trust that this is logically necessary.
+    ({ foo } = await import('./foo.mjs'));
+  });
+
+  it('should do the thing', () => {
+    barMock.mockImplementationOnce(function bar_mock() {/* … */});
+
+    assert.equal(foo(), 42);
+  });
+});
+```
+
+### Services
+
+A little-known fact: node has a builtin way to mock `fetch`. [`undici`](https://github.com/nodejs/undici) is the Node.js implementation of fetch. You do not have to install it—it's shipped with node by default.
```

Comment:
```suggestion
A little-known fact: node has a builtin way to mock `fetch`. [`undici`](https://github.com/nodejs/undici) is the Node.js implementation of `fetch`. You do not have to install it — it's shipped with `node` by default.
```

Reply:
Spaces around an em-dash are not correct